src/dvc_objects/executors.py

```diff
@@ -74,3 +89,56 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         else:
             self.shutdown(wait=True)
         return False
+
+
+async def batch_coros(
+    coros: Sequence[Coroutine],
+    batch_size: Optional[int] = None,
+    callback: Optional[Callback] = None,
+    timeout: Optional[int] = None,
+    return_exceptions: bool = False,
+    nofiles: bool = False,
+) -> List[Any]:
+    """Run the given in coroutines in parallel.
+
+    The asyncio loop will be kept saturated with up to `batch_size` tasks in
+    the loop at a time.
+
+    Tasks are not guaranteed to run in order, but results are returned in the
+    original order.
+    """
```

Comment:
fsspec's `_run_coros_in_chunks` uses `asyncio.gather()` for each batch of tasks. This means that the entire batch of `--jobs/batch_size` tasks must complete before fsspec schedules the next batch to be run.

For operations where each task takes roughly the same amount of time to run this works fine (like when batching simple API calls like `fs.exists`). But for operations where each task's runtime varies (like `fs.put()` and `fs.get()` with multiple files).

If one file in the batch is larger than the others, fsspec waits for that file to be completely uploaded/downloaded before scheduling the next batch of upload/download operations. In the hypothetical scenario that a single upload job is able to completely max out all available bandwidth between my machine and the remote this is ok, but in practice what we actually want is to keep `batch_size` open transfers at the same time, and schedule a new one whenever an existing transfer completes.